core/src/main/java/io/parsingdata/metal/ImmutableObject.java

```diff
@@ -0,0 +1,26 @@
+package io.parsingdata.metal;
+
+/**
+ * When objects are immutable, that means their hashcode will stay the same the moment it is created.
+ * It will improve performance if these objects caches their hash.
+ * <p>
+ * This is a lazy implementation, instead of calculating the hash once within the constructor, to avoid
+ * performance decrease during parsing. In most implementations the hash is not actually used/needed.
+ */
+public abstract class ImmutableObject {
+
+    private Integer hash;
+
+    public abstract int cachingHashCode();
```

Comment:
Maybe this should have a name that more reflects how an implementer should think about it? Such as `fixedHashCode()` or 'initFixedHashCode()` or `permanentHashCode()`?

Maybe `cachingHashCode()` explains that well, I'm just not sure about it because it's also about what it's used for (which is a kind of documentation, but also orthogonal to its functioning?)

Reply:
You could also add some javadoc to explain this more clearly? Not sure, maybe naming is enough.

Reply:
What do you think about `immutableHashCode` as alternative?

> jvdb: Ja, goed, goed!

![Alrighty then!](https://media.tenor.com/Ohu9eiy2s68AAAAC/ace-ventura.gif)